src/external.ts

```diff
@@ -1,42 +1,106 @@
+import { TokenTag } from "./tags";
+import { CurveLpTokenInfo } from "./types";
 
 export const provider = hre.ethers.provider;
-export async function getExternalTokenInfos(
-  chainId: number,
-  underlyingTokenAddresses: string[]
-): Promise<TokenInfo[]> {
-  const baseAssets = underlyingTokenAddresses.map((address) =>
-    ERC20__factory.connect(address, provider)
-  );
 
-  const names = await getTokenNameMulti(baseAssets);
-  const symbols = await getTokenSymbolMulti(baseAssets);
-  const decimals = await getTokenDecimalsMulti(baseAssets);
+const ETH_CONSTANT = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE";
+
+type ExternalTokenInfo = TokenInfo | CurveLpTokenInfo;
 
-  const baseAssetsList = zip(
-    underlyingTokenAddresses,
-    symbols,
-    names,
-    decimals
-  ).map(([address, symbol, name, decimal]): TokenInfo => {
+export async function getExternalTokenInfo(
+  chainId: number,
+  address: string
+): Promise<ExternalTokenInfo> {
+  if (address === ETH_CONSTANT) {
     return {
       chainId,
-      address: address as string,
-      symbol: symbol as string,
-      decimals: decimal as number,
-      name: name as string,
-      // TODO: What logo do we want to show for base assets?
-      // logoURI: ""
+      address,
+      name: "Ether",
+      decimals: 18,
+      symbol: "ETH",
     };
-  });
+  }
+
+  const erc20 = ERC20__factory.connect(address, provider);
+
+  const [name, symbol, decimals] = (await Promise.all(
+    [getTokenName, getTokenSymbol, getTokenDecimals].map((fn) => fn(erc20))
+  )) as [string, string, number];
+
+  const basicTokenInfo = { chainId, address, symbol, name, decimals };
+
+  /* In the case we are building the token list for a non-mainnet network we
+   * should default to returning the simple tokens already specified. The reason
+   * for this is that the Curve.fi ecosystem of contracts is not refletive of
+   * the one on mainnet and in their place we stub with mocked versions. In the
+   * future we could support it on other networks if we wish to build a stubbed
+   * LP system of tokens */
+  if (!name.startsWith("Curve.fi") || chainId !== 1) return basicTokenInfo;
+
+  return await getCurveTokenInfo(basicTokenInfo);
+}
+
+function isCurveLpTokenInfo(
+  tokenInfo: TokenInfo
+): tokenInfo is CurveLpTokenInfo {
+  return !!(tokenInfo.tags && tokenInfo.tags.includes(TokenTag.CURVE));
+}
+
+export async function getExternalTokenInfos(
+  chainId: number,
+  baseTokenAddresses: string[]
+): Promise<ExternalTokenInfo[]> {
+  const baseTokenInfos = await Promise.all(
+    baseTokenAddresses.map((address) => getExternalTokenInfo(chainId, address))
+  );
+
+  const baseSimpleTokenInfos = baseTokenInfos.filter(
+    (info) => !isCurveLpTokenInfo(info)
+  );
+  const baseCurveLpTokenInfos = baseTokenInfos.filter(isCurveLpTokenInfo);
+
+  let externalTokenInfos: ExternalTokenInfo[] = [...baseSimpleTokenInfos];
+  for (const {
+    extensions: { poolAssets },
+  } of baseCurveLpTokenInfos) {
```

Comment:
Can these be forEach instead?

Reply:
No they can't, as we need the processing to be done sequentially, forEach executes async callback functions in paralell as per https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop